Add wp-cli command to delete guest authors, with option to reassign their posts.

smistephen opened the pull request:
Currently, there is no way to delete guest authors from the command line. This command fills that need.

It is activated using `delete-guest-authors`, and passing in the path to a CSV file containing a list of two columns: the ID to delete, and the login to assign the posts to (or "false" to not reassign).

The command runs dry by default, showing the end user the guest authors it would have attempted to delete, and the user login it would have attempted to transfer the posts to.

Passing --no-dry-run, and confirming that you really do want to delete guest authors, actively modifies the database.

Commit:
Merge branch 'master' of https://github.com/Automattic/Co-Authors-Plus

Commit:
Fixed WP CLI create-terms-for-posts if no co-authors found

Script was stopping after no co-authors was find. Now script will
display an error message and continue to the next post.

Commit:
Revert "Fixed WP CLI create-terms-for-posts if no co-authors found"

This reverts commit f75974429b8b804ea9e49f59f969aacd2eda1ddb.

Commit:
Revert "Removed duplicate left join for optimization"

This reverts commit 6e7da2bd54b5b7b8f87e4c5226745c13a7506a11.

Commit:
Fixed WP CLI create-terms-for-posts if no co-authors found

Script was stopping after no co-authors was find. Now script will
display an error message and continue to the next post.

Commit:
Removed double left join on posts_join_filter (#419)

Resolves #417

* Removed duplicate left join for optimization

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
CoAuthors display now for Pages and JS is being enqueued on Pages

Commit:
JS error goes away from js_vars()

Commit:
changed is_post_type_enabled() to clean up code

Commit:
removed extra whitespace

Commit:
removed all traces of $post_type in functions, since is_post_type_enabled() should do the trick now

trepmal committed:
Revert "Update version constant"

This reverts commit 525f211c38b56e6b52bbb1e8b88e8f408dc411be.

Commit:
Edited words to match glossary

Commit:
Removed continue because the script would stop running.

Commit:
Added better consistencies with glossary

trepmal committed:
Merge branch 'master' into reduce-sleep

trepmal committed:
Revert "Revert "Update version constant""

This reverts commit 5ac7d9a2288d49db6462a98d9211dfa42480a0f7.

Commit:
doesn't return false after quick edit saves

Commit:
replaced hardcoded 'author' with $this->coauthor_taxonomy

philipjohn committed:
Tested tag to 4.8

philipjohn committed:
Merge pull request #400 from Automattic/reduce-sleep

Reduce amount of sleep

philipjohn committed:
Merge pull request #420 from rebeccahum/cli-create-terms-for-posts

Fixed WP CLI create-terms-for-posts if no co-authors found

philipjohn committed:
Merge pull request #423 from rebeccahum/terminology

Terminology updated throughout

philipjohn committed:
Merge pull request #426 from rebeccahum/replace_hardcoded_author

Replace hardcoded 'author' with $this->$coauthor_taxonomy

trepmal committed:
Move parenthesis to fix esc_html and sprintf

Fixes warning: `Warning: sprintf(): Too few arguments in /path/wp-content/plugins/co-authors-plus/php/class-coauthors-guest-authors.php on line 487`

Warning surfaces when deleting a guest author that is mapped to a WP user

trepmal committed:
Move parenthesis to fix esc_html and sprintf

Fixes warning: `Warning: sprintf(): Too few arguments in /path/wp-content/plugins/co-authors-plus/php/class-coauthors-guest-authors.php on line 487`

Warning surfaces when deleting a guest author that is mapped to a WP user

trepmal committed:
Move parenthesis to fix esc_html and sprintf

Fixes warning: `Warning: sprintf(): Too few arguments in /path/wp-content/plugins/co-authors-plus/php/class-coauthors-guest-authors.php on line 487`

Warning surfaces when deleting a guest author that is mapped to a WP user

Commit:
Added progress so users have an idea of how long it will take

philipjohn committed:
Merge pull request #430 from Automattic/trepmal-patch-1

Move parenthesis to fix esc_html and sprintf

Commit:
Added a better style of progress bar!

Commit:
Deleting guest authors is less confusing

Commit:
Fixed merge conflicts

Commit:
Merge remote-tracking branch 'refs/remotes/Automattic/master'

Commit:
Hopefully now merge conflict is gone now...

Commit:
some extra notes.

Commit:
Merge remote-tracking branch 'refs/remotes/Automattic/master' into delete_guest_author

# Conflicts:
#	php/class-coauthors-guest-authors.php

Commit:
featured image to avatar

Commit:
featured image to avatar

Commit:
featured image to avatar

Commit:
Removed avatar sizing

Commit:
Removed avatar sizing

Commit:
Removed avatar sizing

Commit:
remove duplicated linked author byline

Commit:
remove duplicated linked author byline

Commit:
remove duplicated linked author byline

Commit:
coauthors_wp_list_authors has the option now of listing only guest authors

philipjohn committed:
Merge pull request #433 from rebeccahum/featured_image_avatar

Guest author's featured image is avatar now

philipjohn committed:
Merge pull request #434 from rebeccahum/remove_avatar_sizing

Removed extra image sizing

philipjohn committed:
Merge pull request #435 from rebeccahum/remove_duplicated_byline

Remove duplicated byline

Commit:
remove duplicates from linked accounts

Commit:
changes as per philipjoin

Commit:
Added later escaping

Commit:
Merge remote-tracking branch 'refs/remotes/Automattic/master'

Commit:
Use linked account for accurate post count

Commit:
Redid README.md with better markdown formatting

Commit:
Redid README.md with better markdown formatting

Commit:
added readme.txt back

Commit:
added readme.txt back

Commit:
remove quotes, stray =s and upgrade notice section

Commit:
remove quotes, stray =s and upgrade notice section

Commit:
added check for false $term

trepmal committed:
Do not 404 when guest author archive has 0 posts. Addresses #425 (#440)

Resolves #425, #409

trepmal committed:
Do not 404 when guest author archive has 0 posts. Addresses #425 (#440)

Resolves #425, #409

Commit:
Merge remote-tracking branch 'refs/remotes/Automattic/master'

Commit:
Merge branch 'master' of https://github.com/rebeccahum/Co-Authors-Plus

Commit:
Merge remote-tracking branch 'refs/remotes/Automattic/master'

Commit:
fixed author archive title

Commit:
revised comment

jordie23 committed:
Fix invalid CSS (#442)

jordie23 committed:
Fix invalid CSS (#442)

Commit:
fixed coauthors_link_single to work with guest authors

Commit:
added _blank target

Commit:
switched logic around

Commit:
Merge remote-tracking branch 'refs/remotes/Automattic/master'

Commit:
added hooks for creating/deleting guest authors

Commit:
added hooks for creating/deleting guest authors

sboisvert committed:
Merge pull request #446 from rebeccahum/added_guest-author_hooks

Added guest author hooks for create/delete

The hooks are useful in being able to manage caches that take into account guest authors.

Commit:
Merge remote-tracking branch 'refs/remotes/Automattic/master'

Commit:
Merge pull request #439 from rebeccahum/new_readme.MD

New README.md

Commit:
Merge remote-tracking branch 'refs/remotes/Automattic/master'

sboisvert committed:
Merge pull request #438 from rebeccahum/accurate_guest_post_count

Accurate Guest Author post count on linked accounts

Commit:
fix logic for doing_autosave check

Commit:
Merge remote-tracking branch 'refs/remotes/Automattic/master'

Commit:
Merge branch 'master' of https://github.com/rebeccahum/Co-Authors-Plus

Commit:
Merge remote-tracking branch 'refs/remotes/origin/user_login-spaces' into upstream/master

Commit:
Fixes "Notice: Trying to get property of non-object" for users who have user_login with spaces

Commit:
Fixes issue of not being able to reassign posts to user with user_login with spaces on deletion of a user

philipjohn committed:
Merge pull request #437 from rebeccahum/dedupe_coauthors_wp_list_authors

remove duplicates from linked accounts on coauthors_wp_list_authors()

philipjohn committed:
Merge pull request #436 from rebeccahum/coauthors_wp_list_authors

coauthors_wp_list_authors() has option to list only guest authors now

philipjohn committed:
Merge pull request #432 from rebeccahum/delete_guest_author

Deleting guest authors is less confusing

philipjohn committed:
Merge pull request #407 from haleeben/fix/no_authors_found_message

Author search UX issues

philipjohn committed:
Merge pull request #441 from rebeccahum/filter_author_archive

Filter author archive

philipjohn committed:
Merge pull request #414 from mdchiragpatel/fixed-120

Remove associated guest user when mapped user id deleted.

philipjohn committed:
Merge pull request #453 from rebeccahum/user_login_spaces

user_login spaces problem when using add_coauthors

Commit:
Adding escaping

Commit:
Merge remote-tracking branch 'refs/remotes/Automattic/master'

pkevan committed:
Adding details of filter for slow performance

- Use simple tax queries for CAP to improve performance
- Detail of filter to add to theme
- Adding CLI command to run

philipjohn committed:
Merge pull request #444 from rebeccahum/fix_coauthors_links_single

Fix coauthors_links_single()

philipjohn committed:
Merge pull request #456 from Automattic/fix/issue319

Adding details of filter for slow performance

philipjohn committed:
Remove redundant test for 404 on Author Archive

Since we made CAP consistent with core in it's handling of author archives where there are no posts (see https://github.com/Automattic/Co-Authors-Plus/issues/425) we no longer expect there to be a 404 response on the author archive for a guest author.

Commit:
Merge remote-tracking branch 'refs/remotes/Automattic/master'

Commit:
Fixed counting for guest author posts

Commit:
Users table will give accurate post count if linked guest author has post(s)

Commit:
No need to add $guest_count->count anymore now that the Users table displays the count properly

Commit:
Added one more validation

sboisvert committed:
Merge pull request #461 from rebeccahum/guest_author_count_table

Guest Author Counts are more accurate

sboisvert committed:
Merge pull request #450 from mslinnea/fix-autosave-logic

Fixes logic for DOING_AUTOSAVE check

sboisvert committed:
Merge pull request #422 from rebeccahum/pages_quick_edit

Pages archive now displays coauthors and quick edit works

sboisvert committed:
Merge pull request #386 from trepmal/milestone/kl-375

Fix private post viewing on front-end

rebeccahum committed:
Merge pull request #431 from rebeccahum/wpcli_create-guest-author_progress

Added progress to create-guest-authors so users have an idea of how long it will take

Commit:
Merge remote-tracking branch 'refs/remotes/Automattic/master'

Commit:
set $coauthors_loading

rebeccahum committed:
Merge pull request #468 from rebeccahum/set-coauthors_loading

Set $coauthors_loading

sanketio committed:
[#198] Fix guest authors with non-ASCII characters

sanketio committed:
[#198] Replace .html() with .text() for simple string in js

sanketio committed:
[#279] Fix duplicate name display issue

 - When `coauthors_auto_apply_template_tags` is set to true, `coauthors_posts_links()` or `the_author_posts_link()` methods displays duplicate names if post has multiple authors.

sanketio committed:
[#198] Improve unit testing for non-ASCII character issue

sanketio committed:
[#279] Add issue link as ref

sanketio committed:
[#279] Improve phpunit test cases

sanketio committed:
[#198] Improve phpunit test cases + cosmetic changes

sanketio committed:
[#198] Assign values to variables instead of keeping empty for test cases

sanketio committed:
[#279] Improve condition checking for unit test cases

sanketio committed:
[#279] Refactor comments for test cases

sanketio committed:
[#184] Started unit tests for template tags

- Added unit tests for get_coauthors() & is_coauthor_for_post()

sanketio committed:
[#184] Unit tests for coauthors() for template tags

sanketio committed:
[#184] Unit tests for coauthors_posts_links_single() for template tags

sanketio committed:
[#184] Unit tests for coauthors_firstnames() for template tags

sanketio committed:
[#184] Method description for coauthors_firstnames() for template tags

sanketio committed:
[#184] Unit tests for coauthors_lastnames() for template tags

sanketio committed:
[#184] Unit tests for coauthors_nicknames() for template tags

sanketio committed:
[#184] Unit tests for coauthors_emails() for template tags

sanketio committed:
[#184] Unit tests for coauthors_links_single() for template tags

sanketio committed:
[#184] Unit tests for coauthors_ids() for template tags

sanketio committed:
[#184] Unit tests for get_the_coauthor_meta() for template tags

sanketio committed:
[#184] Unit tests for coauthors_wp_list_authors() for template tags

sanketio committed:
[#184] Unit tests for coauthors_get_avatar() for template tags

sanketio committed:
[#184] Fix undefined variable warnings and add feed_type for feed link

sanketio committed:
[#388] Unit tests for is_guest_authors_enabled() method

sanketio committed:
[#388] Unit tests for get_coauthor_by() method

sanketio committed:
[#184] Improve unit tests

sanketio committed:
[#388] Unit tests for is_post_type_enabled() method

sanketio committed:
[#388] Fix unit tests for is_post_type_enabled() method using current screen

sanketio committed:
[#184] Remove not needed assertions and add proper commenting

sanketio committed:
[#388] Unit tests for current_user_can_set_authors() method

- Also fixed warning "Trying to get property of non-object" if post_type is not set in "get_current_screen()" on https://github.com/Automattic/Co-Authors-Plus/blob/master/co-authors-plus.php#L1014

sanketio committed:
[#388] Unit tests for search_authors() method

sanketio committed:
[#388] Unit tests for get_author_term() method

sanketio committed:
[#388] Unit tests for update_author_term() method

sanketio committed:
[#388] Unit tests for get_guest_author_by() method

sanketio committed:
[#388] Unit tests for get_guest_author_thumbnail() method

sanketio committed:
[#388] Unit tests for get_guest_author_fields() method

sanketio committed:
[#388] Unit tests for get_all_linked_accounts() method

sanketio committed:
[#388] Unit tests for create_guest_author_from_user_id() method

sanketio committed:
[#388] Unit tests for handle_delete_guest_author_action() method

sanketio committed:
[#388] Unit tests for CoAuthors_Guest_Authors::delete() method

sanketio committed:
[#388] Unit tests for coauthors orders

sanketio committed:
[#388] Improve unit tests

sanketio committed:
[#184] Unit tests for coauthors orders

sanketio committed:
[#388] Remove unit tests for coauthors orders

sanketio committed:
[#388] Improve unit tests

megfh committed:
Update class-wp-cli.php

fix #472

rebeccahum committed:
Merge pull request #480 from megfh/patch-1

Update class-wp-cli.php

sanketio committed:
[#198] Replace urlencode/urldecode with rawurlencode/rawurldecode

sanketio committed:
[#279] Code cleanup

- Convert duplicate conditions into variables and use those variables.

sanketio committed:
[#388] Modify where condition when WordPress generates `post_author IN (id)` in query

philipjohn committed:
Merge pull request #473 from sanketio/fix/198

Fix the issue where guest authors with non-ASCII characters can't be used as co-authors

sanketio committed:
[#458] Update travis file

nmeh20 committed:
Changes to resolve issue #332 about missing coauthor meta in get_the_coauthor_meta function

nmeh20 committed:
Changes to resolve issue #332. Updated funtion resolve error generated while fetching 'type' field.

nmeh20 committed:
Changes for optimizing code of get_the_coauthor_meta function

sanketio committed:
[#458] Include WP 4.6 env in travis file

philipjohn committed:
Merge pull request #482 from sanketio/fix/458

Update .travis.yml file for PHPUnit tests

philipjohn committed:
Merge pull request #476 from sanketio/fix/388

Review and improve test coverage

philipjohn committed:
Merge pull request #475 from sanketio/fix/184

Unit tests/Fix warnings for template tags

philipjohn committed:
Merge pull request #457 from Automattic/fix/author_404_test

Remove redundant test for 404 on Author Archive

nmeh20 committed:
Updating function to allow fetching meta data based on user_id to support the_coauthor_meta function

nmeh20 committed:
Removing unused global variables

sanketio committed:
Merge branch 'master' into fix/279

# Conflicts:
#	tests/test-template-tags.php

nmeh20 committed:
Formatting in my changes as per PHPCS

nmeh20 committed:
Space alignment and adding escape to output

philipjohn committed:
Merge pull request #474 from sanketio/fix/279

Fix the issue where incompatibility when `coauthors_auto_apply_template_tags` set to true

philipjohn committed:
Release v3.2: Update version numbers etc

philipjohn committed:
Merge pull request #484 from nilzari/master

Changes to resolve issue #332 about missing coauthor meta

philipjohn committed:
Release v3.3: Changelog and props

sanketio committed:
Fix failing test cases

philipjohn committed:
Merge pull request #500 from sanketio/fix/499

Fix failing test cases for coauthors functions

philipjohn committed:
Merge branch 'release/3.3'

binodkalathil committed:
Unexpected output of template tag 'coauthors_links' is resolved by updating conditions

The template tag was outputting usernames of guest authors when they do not have a website set in their profile. It will now output guest authors display names as expected.

binodkalathil committed:
Corrected PHPUnit test cases to work with the udpated code

A couple of test cases were failing as the return value of a function has changed as per the modifications done for fixing the issue(Issue #469)

pkevan committed:
Bump last version of WP to 4.8

binodkalathil committed:
Merge branch 'master' into issues/469-function-return-author-login-name-instead-of-display-name

sboisvert committed:
Merge pull request #504 from binodkalathil/issues/469-function-return-author-login-name-instead-of-display-name

Unexpected output of template tag 'coauthors_links' is resolved by updating conditions

rebeccahum committed:
Revert "Merge pull request #438 from rebeccahum/accurate_guest_post_count"

This reverts commit 2953a4cc26ab9d1d46ad25b726b355ab00e39943, reversing
changes made to 32c7f09e4aac73c9b59962c21f3a02b4640e96f4.

blunce24 committed:
Added i18n support to Author title and name for author archive

blunce24 committed:
changed author display name on author archive page to be Full Name rather than slug

blunce24 committed:
added check for  before using it in query

blunce24 committed:
updated truthiness check for current_coauthor_term object

blunce24 committed:
formatting adjustments

blunce24 committed:
formatting updates

blunce24 committed:
allow PHPCS inspection of file

TheCrowned committed:
Replacing search_fields argument (non-existent) in get_users() call with search_columns.

rebeccahum committed:
Merge pull request #516 from blunce24/i18n-author-archive

Added i18n support and corrected author display name for author archive: addressing issue #515

blunce24 committed:
fixed tab issue

blunce24 committed:
changed check on coauthor term to be more concise

sboisvert committed:
Merge pull request #517 from blunce24/current-user-no-terms-warning

check for coauthor terms before using them in query: addressing issue #511

philipjohn committed:
Merge pull request #519 from TheCrowned/fix/513-author-search-argument

Replacing search_fields with search_columns in get_users()

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Add isset() checkes to post registration

Commit:
remove trailing whitespace

TheCrowned committed:
Updated readme with FAQ on how to disable Guest Authors. Fix #122.

mdbitz committed:
Merge pull request #521 from paulschreiber/fix/warnings

Adds isset() checks to post type registration

TheCrowned committed:
Fix test_coauthors_links_single_when_guets_author(), which was wrongly passing.

The test passed because was updating a postmeta instead of the user record in wp_users, thus subsequent assertions all passed simply because no real edit had been done.
When making the edit for real, the test would fail in several places, so tweaks were needed to fix its behavior.

TheCrowned committed:
Updated coauthors_links_single so that CAP coauthor link matches Core author link structure.

TheCrowned committed:
Removed function action_pre_user_query and its hooks, as it was obsolete

TheCrowned committed:
Fix possible discrepancy in author search due to mismatched use of user_login and user_nicename.

sboisvert committed:
Merge pull request #529 from TheCrowned/fix/505-test-wrong-meta-key

Test test_coauthors_links_singe_guests_authors wrongly passing

sboisvert committed:
Merge pull request #523 from TheCrowned/fix/122-faq-disable-guest-author

Added FAQ to readme on how to disable Guest Authors

jacobarriola committed:
Only filter author template for title

Repro steps:
- in `archive.php`, use `the_archive_title()` method
- load a category archive page

Expected results:
"Category: category_name"

Actual results:
null

Notes:
The `filter_author_archive_title()` function is hooked to the `get_the_archive_title` filter, which can run in other archive templates such as `category` or `tag`. When run in such a template, the function does not return the `$title` parameter, resulting in `null`.

sboisvert committed:
Merge pull request #531 from TheCrowned/fix/530-remove-action_pre_user_query

Removed function action_pre_user_query and its hooks

TheCrowned committed:
Removed whole setUp function, which contained the two filters that made WP queries non-TEMPORARY

TheCrowned committed:
Skipping irrelevant cap checks through list of allowed caps.

TheCrowned committed:
Merge branch 'master' of https://github.com/Automattic/Co-Authors-Plus into fix/489-skip-filtering-for-irrelevant-caps

mariovalney committed:
Allow guest authors have personal data exported (#528)

* [Style] Identation changed to tab

* [Feat] Added personal data exporter to guest authors

* [Feat] Added a filter to allow third part plugins add guest author data

* [Style] WPCS fixes

* [Fix] Filter name changed

* [Style] WPCS fixes

* [Fix] Filter name changed

TheCrowned committed:
Updated tests to search_authors() to use assertArrayNotHasKey instead of assertNotContains

TheCrowned committed:
Re-enabled Test_Author_Queries->test__author_arg__user_is_coauthor__author_arg()

TheCrowned committed:
Re-enabled Test_Author_Queries->tests__author_name_arg_plus_tax_query__is_coauthor()

TheCrowned committed:
Improved inline doc

sboisvert committed:
Merge pull request #546 from TheCrowned/fix/538-re-enable-tests

Re-enable disabled Test_Author_Queries tests

Commit:
Add post type parameter to the Mine link, so it works for Pages

rebeccahum committed:
Merge pull request #547 from pdemarte/post-type-pages-fix

Add post type parameter to the Mine link, so it works for Pages

jacobarriola committed:
Merge branch 'master' into patch-1

rebeccahum committed:
Merge pull request #535 from jacobarriola/patch-1

🐛Only filter author template for title

rebeccahum committed:
Merge pull request #532 from TheCrowned/fix/273-search-discrepancy

Fix possible discrepancy in author search with ignored authors

rebeccahum committed:
Merge pull request #543 from TheCrowned/fix/489-skip-filtering-for-irrelevant-caps

Skip irrelevant capabilities filtering in filter_user_has_cap() with whitelist

TheCrowned committed:
Changed get_user_by('slug' to get_user_by('login' when using linked_account

mdbitz committed:
Merge pull request #540 from TheCrowned/fix/526-non-temporary-queries-on-multisite-test

Removed Test_Author_Queried_Object setUp function

mdbitz committed:
Merge pull request #558 from TheCrowned/fix/554-user-login-linked-account

column_posts() should use user_login when querying for linked_account

TheCrowned committed:
Only using new `cap-` prefixed term when counting user posts.

sboisvert committed:
Merge pull request #567 from TheCrowned/fix/561-wrong-post-count-guest-authors

Fixing double posts counts for users with linked accounts

TheCrowned committed:
$ignored_authors in search should use user_nicename, not login.

rebeccahum committed:
Merge pull request #569 from TheCrowned/fix/568-failing-test

$ignored_authors in search should use user_nicename, not login.

smistephen committed:
Add wp-cli command to delete guest authors, with option to reassign their posts.

Currently, there is no way to delete guest authors from the command line. This command fills that need.

It is activated using `delete-guest-authors`, and passing in the path to a CSV file containing a list of two columns: the ID to delete, and the login to assign the posts to (or "false" to not reassign).

The command runs dry by default, showing the end user the guest authors it would have attempted to delete, and the user login it would have attempted to transfer the posts to.

Passing --no-dry-run, and confirming that you really do want to delete guest authors, actively modifies the database.

smistephen commented:
I'm not sure why the Travis build fails, it passes in my fork:

https://github.com/smistephen/Co-Authors-Plus/runs/12256989

TheCrowned commented:
This looks great! Could we allow deletion of all guest authors that have linked accounts? See #550. Would be good for testing, because allows to basically revert the install process. Maybe the csv thing could become an option (not compulsory) and in some way incorporate this bulk automatic deletion. Only include guest authors with linked accounts as we wouldn't want to delete the others, as there would be no fallback in post assignment.

smistephen commented:
Totally :+1:

Could you open a separate issue for those improvements? This PR's already pretty big, haha.

TheCrowned commented:
It's already in #550

smistephen commented:
Perfect

smistephen commented:
Looking at #550, it seems like it could be its own independent PR. Am I reading it right?

sboisvert reviewed (commented):
A few questions :)

smistephen commented on php/class-wp-cli.php:
```diff
@@ -45,6 +45,120 @@ public function create_guest_authors( $args, $assoc_args ) {
+					$assignee = false;
+				}
+
+				$delete_status = $coauthors_plus->guest_authors->delete( $id, $assignee );
```
The function returns a WP_Error:
https://github.com/Automattic/Co-Authors-Plus/blob/master/php/class-coauthors-guest-authors.php#L1304

sboisvert replied:
Would that error message be enough to debug the problem? Is there more information we could add to help the end user?

smistephen commented on php/class-wp-cli.php:
```diff
@@ -45,6 +45,120 @@ public function create_guest_authors( $args, $assoc_args ) {
+
+		$filename = $args[0];
+
+		// Add custom error handling on fopen error.
```
I felt it was a clearer variable name.

sboisvert replied:
Sorry, I meant the custom error handling, not the variable name :)

smistephen replied:
Ah. The custom error handling is there to avoid an ugly PHP warning and stack trace from being thrown to the console if the CSV file can't be opened, doesn't exist, or some other error.

smistephen commented on php/class-wp-cli.php:
```diff
@@ -45,6 +45,120 @@ public function create_guest_authors( $args, $assoc_args ) {
+			if ( true === $dry_run ) {
+				WP_CLI::log( sprintf(
+					/* translators: 1: Guest author ID 2: Intended assignee */
+					__( 'Attempting to delete guest author %1$d, intended assignee %2$s', 'co-authors-plus' ),
```
Sure, I'll see what I can pluck out :+1:

sboisvert commented on php/class-wp-cli.php:
```diff
@@ -45,6 +45,120 @@ public function create_guest_authors( $args, $assoc_args ) {
+				} else {
+					WP_CLI::log( sprintf(
+						/* translators: %d: Guest author ID */
+						__( 'Successfully deleted guest author %d', 'co-authors-plus' ),
```
If someone looked back at this log file file. Is there more information they would need?

smistephen committed:
Add additional log output.

This commit adds additional information on success and failure, telling the end user not only the ID of the guest author that was deleted, but also (if applicable) the login their posts were assigned to.

sboisvert commented:
Nod, it might be useful for the comment to explain the intent behind this
since I wouldn't say it's a common pattern :)

On 14 August 2018 at 18:40, Stephen Smith <notifications@github.com> wrote:

> *@smistephen* commented on this pull request.
> ------------------------------
>
> In php/class-wp-cli.php
> <https://github.com/Automattic/Co-Authors-Plus/pull/588#discussion_r210126512>
> :
>
> > +		);
> +
> +		$assoc_args = wp_parse_args( $assoc_args, $defaults );
> +
> +		$dry_run = $assoc_args['dry-run'];
> +
> +		// Safety first.
> +		$filename_status = validate_file( $args[0] );
> +
> +		if ( 0 !== $filename_status ) {
> +			WP_CLI::error( __( 'The provided file cannot be accessed. Please use either relative or absolute pathnames, and avoid using directory traversal such as ../ and other methods.', 'co-authors-plus' ) );
> +		}
> +
> +		$filename = $args[0];
> +
> +		// Add custom error handling on fopen error.
>
> Ah. The custom error handling is there to avoid an ugly PHP warning and
> stack trace from being thrown to the console if the CSV file can't be
> opened, doesn't exist, or some other error.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/Automattic/Co-Authors-Plus/pull/588#discussion_r210126512>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AARpqYPBy7Pw6eRbB4T6uFU6dxWvrJgzks5uQ1HagaJpZM4V9Lsf>
> .
>

smistephen committed:
Add more information to dry run output.

This commit utilizes the CAP API during dry runs to pull out information about the guest authors about to be deleted, to make sure they're the right ones.

smistephen committed:
Add more information to command output.

This commit utilizes the CAP API to pull out information about the guest authors about to be deleted, to make sure they're the right ones.

smistephen committed:
Add additional explanatory comments about the custom error handling code

philipjohn commented on php/class-wp-cli.php:
```diff
@@ -45,6 +45,145 @@ public function create_guest_authors( $args, $assoc_args ) {
+	 * ## OPTIONS
+	 *
+	 * <guest-author-ids-file>
+	 * : Provide the location of a CSV file containing a list of two columns: guest author ID to delete, and either the string "false" (which means the deleted guest author's posts will not be reassigned) or a string denoting the login of the user or guest author the deleted guest author's posts are to be reassigned to. No column headers are required or allowed, and the maximum line length including ...
```
Could we break this into multiple lines for easier readability?

philipjohn commented on php/class-wp-cli.php:
```diff
@@ -45,6 +45,145 @@ public function create_guest_authors( $args, $assoc_args ) {
+				}
+			} else {
+				// Have to cast so the function won't literally try and reassign to user "false".
+				if ( 'false' === $assignee ) {
```
What happens if someone has the username "false"? Or "null" or similar?

smistephen replied:
Haha, you tricky trickster. I've added code to deal with that now :+1:

philipjohn commented on php/class-wp-cli.php:
```diff
@@ -45,6 +45,145 @@ public function create_guest_authors( $args, $assoc_args ) {
+
+		while ( ( $line = fgetcsv( $file, 100, ',' ) ) !== false ) {
+			$id       = filter_var( $line[0], FILTER_SANITIZE_NUMBER_INT );
+			$assignee = filter_var( $line[1], FILTER_SANITIZE_STRING );
```
What happens if this assignee doesn't exist?

smistephen replied:
The function returns a WP_Error :+1: 
https://github.com/Automattic/Co-Authors-Plus/pull/588#discussion_r210121608

philipjohn replied:
I was referring to the value of `$assignee` - what happens if the assignee (i.e. the username) given in the CSV doesn't exist? What will this code do to that value?

smistephen replied:
Ah yes. That is an error, and it has now been fixed.

philipjohn reviewed (commented):
This is looking good - I've left some comment.

One overall question - how can we check this code is going to work as expected in production?

smistephen commented:
> One overall question - how can we check this code is going to work as expected in production?

Checks I've performed so far:
 - run phpunit _(passes)_
 - run phpcs _(passes)_
 - look for performance problems and security vulnerabilities _(nothing obvious)_
 - test function return values _(added appropriate error handling code)_
 - test on a sample WordPress instance _(seems to work as promised)_
 - try to break it with faulty data _(seems to fail gracefully as predicted)_

And I'm always open to more opportunities :+1:

smistephen committed:
regenerate pot file

Allow for remapping to user named "false".

The CAP delete function takes, as its second parameter, either a string containing the login of the user you wish to reassign the deleted guest author's posts to, or a boolean false stating that you do wish to reassign.

But what if you want to reassign to a user named false? Before that was not possible.

Now, the code specifically looks for if the instance has a user named false, and if so, allows it to be reassigned to. If not, it assumes the boolean was intended.

smistephen committed:
Break documentation comment onto multiple lines for readability

smistephen commented:
Hooray for @philipjohn pulling the gremlin out of Travis, builds pass now, as they should :+1:

philipjohn commented:
> run phpunit (passes)

Which unit tests did you look at when trying to assess whether this code works as intended?

> try to break it with faulty data (seems to fail gracefully as predicted)

That's a good idea! What kind of faulty data did you use here?

smistephen committed:
Add existence checking for assignee value

smistephen committed:
Rearrange dry run command output for greater clarity

smistephen commented on php/class-wp-cli.php:
```diff
@@ -45,6 +45,145 @@ public function create_guest_authors( $args, $assoc_args ) {
+				} else {
+					WP_CLI::log( sprintf(
+						/* translators: 1: Guest author ID 2: Intended assignee 3: Guest author display name 4: Guest author email 5: Guest author linked account */
+						__( 'Attempting to delete guest author %1$d, intended assignee %2$s. Display Name: %3$s Email: %4$s Linked Account: %5$s', 'co-authors-plus' ),
```
I've rearranged the output, similar to elsewhere in the file: https://github.com/Automattic/Co-Authors-Plus/blob/78ce17c2b820f8baa1f35254c0f03b75184e1755/php/class-wp-cli.php#L162

chrissy-dev commented:
@rebeccahum I appreciate this is an old PR but is there a reason this PR never got merged? Happy to help get this over the line if possible.

rebeccahum commented:
@chrissy-dev I think this just slipped the radar. First step would be to resolve any merge conflicts and test if the current PR still works please!

GaryJones commented:
Overlaps with #696.
